### PR TITLE
Parse fdinfo and inotify stats from /proc/[pid]/fdinfo/[fd]

### PR DESCRIPTION
**Pull request by pirxthepilot**
First attempt at adding functionality to support #107. Not quite sure if my approach here is correct - guidance/feedback would be very much appreciated. Also thought I'd let you guys review first before I write the tests.

Thanks!

### CODE REVIEW

File: inotify.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+package procfs
+
+import (
+	"bufio"
+	"os"
+	"regexp"
+	"strings"
+)
+
```
Comment (pgier) [inline inotify.go]: I think it would be good to create an additional struct here called `FDInfo` which contains pos, flags, and mnt_id since those fields appear to be consistently available.  The `FDInfo` struct could contain a slice of `InotifyStat` containing the detailed information for inotify file descriptors.
And to be more consistent with the rest of the API, the file should be called `proc_fdinfo.go`.

File: inotify.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// InotifyStat represents a file descriptor's inotify watch count.
+type InotifyStat struct {
+	// File descriptor
+	FD string
```
Comment (pgier) [inline inotify.go]: This could be moved into the `FDInfo` struct.

File: inotify.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	// List of inotify lines in the fdinfo file
+	Lines []string
```
Comment (pgier) [inline inotify.go]: Instead of just loading the line here, I think it would be better to parse the values into their expected fields/types.  Or at least add the values to map for easy lookup.

File: proc.go [hunk 4]
```diff
@@ -256,3 +256,18 @@ func (p Proc) fileDescriptors() ([]string, error) {
 func (p Proc) path(pa ...string) string {
 	return p.fs.Path(append([]string{strconv.Itoa(p.PID)}, pa...)...)
 }
+
+func (p Proc) fileDescriptorInfo(fd string) ([]byte, error) {
```
Comment (pgier) [inline proc.go]: This needs to capitalized to be available from outside of the procfs package.  Also, I think this should return a `FDInfo` with the data already parsed, to make it easier to use.

**Comment by discordianfish** [comment]
Agree with pgier and don't have more to add. And sorry for not responding earlier!

**Comment by pirxthepilot** [comment]
Hi folks, just pushed some updates based on your feedback. I replaced `fileDescriptorInfo` with `FileDescriptorsInfo`, which now returns a slice of `FDInfo` types. Also modified and moved `InotifyWatchLen` to `proc.go` as it felt more correct to me.

I'm sure this still needs some tweaks. Also, my go-foo isn't too good to begin with :) Hope it at least makes some sense!

**Comment by pgier** [comment]
@pirxthepilot Looks good, but would you mind rebasing on latest master?  I think the staticcheck CI failures are related to an outdated Makefile or some other build config.

**Comment by pirxthepilot** [comment]
@pgier whoops, sorry bout that! done!

File: proc_fdinfo.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	// File access mode and status flags
+	Flags string
+	// Mount point ID
+	MntID string
+	// List of inotify lines (structed) in the fdinfo file (kernel 3.8+ only)
+	InotifyInfos []InotifyInfo
+}
+
+// FDInfo constructor. On kernels older than 3.8, InotifyInfos will always be empty.
+func (p Proc) newFDInfo(fd string) (*FDInfo, error) {
```
Comment (pgier) [inline proc_fdinfo.go]: Can you change this to just `FDInfo`?  We did some refactoring recently to remove `new` from the functions where it's not needed.

Comment (pirxthepilot) [reply proc_fdinfo.go]: Will try to have this done over the weekend. Do you mean just outright rename this function to `FDInfo`, i.e. 
```
func (p Proc) FDInfo(fd string) ...
```
And should I do the same for `newInotifyInfo` (further down this file)?

Thanks!

Comment (pgier) [reply proc_fdinfo.go]: Right, just rename to `func (p Proc) FDInfo(fd string)`.  You can take a look at some of the other `proc_<thing>.go` files for example.  
I'm thinking `newInotifyInfo` should be `parseInotifyInfo` instead.  It can't be just `InotifyInfo` because then it would conflict with the struct name.

**Comment by pgier** [comment]
@pirxthepilot I think this still needs unit tests before we merge.

**Comment by pirxthepilot** [comment]
@pgier sorry for the delayed response. I pushed a new commit with the requested changes. Just want to run it by you as I start writing the tests.

As for the tests, how should I go about adding fixtures? Do I just append data by hand in `fixtures.ttar`?

File: proc.go [hunk 6]
```diff
@@ -265,3 +265,37 @@ func (p Proc) fileDescriptors() ([]string, error) {
...
+		fdinfo, _ := p.FDInfo(n)
+		fdinfos[i] = *fdinfo
+	}
+
+	return fdinfos, nil
+}
+
+// InotifyWatchLen returns the total number of inotify watches used
+// by the process.
+func (p Proc) InotifyWatchLen() (int, error) {
```
Comment (pgier) [inline proc.go]: It might be better to make this a regular function which takes a slice of `ProcFDInfo` as an argument, instead of a method of Proc.  Since the process is not used other than to get the FDInfos.  Or you could define a new type which is a `[]ProcFDInfo` and make this a method of that type.

File: proc_fdinfo.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 The Prometheus Authors
```
Comment (pgier) [inline proc_fdinfo.go]: Should be 2019

**Comment by pirxthepilot** [comment]
Changes made and tests written. I think the test in `proc_test.go` should provide coverage for the functions there, except for `InotifyWatchLen`, which I wrote a separate test for in `proc_fdinfo_test.go`. Let me know if you need anything else.

File: proc.go [hunk 8]
```diff
@@ -279,3 +279,21 @@ func (p Proc) fileDescriptors() ([]string, error) {
...
+	names, err := p.fileDescriptors()
+	if err != nil {
+		return nil, err
+	}
+
+	fdinfos := make(ProcFDInfos, len(names))
+
+	for i, n := range names {
+		fdinfo, _ := p.FDInfo(n)
+		fdinfos[i] = *fdinfo
```
Comment (pgier) [inline proc.go]: If `p.FDInfo` returns an error and a nil fdinfo won't this cause a nil pointer dereference?  It seems like this could happen for example if a file is closed while iterating through the files.  I think it would be better to check for an error and then use `append` to add to the slice instead of using the index.

Also, if you make `ProcFDInfos` a slice of pointers instead of a slice of structs, then it should be a bit more efficient because the full struct don't need to be copied with each assignment.

File: fixtures.ttar [hunk 9]
```diff
@@ -2079,70 +2147,78 @@ Mode: 755
...
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
-Directory: fixtures/sys/kernel/config/target/iscsi/iqn.2003-01.org.linux-iscsi.osd1.x8664:sn.abcd1abcd2ab/tpgt_1
+Directory: fixtures/sys/kernel/config/target/core/fileio_1
 Mode: 755
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
-Path: fixtures/sys/kernel/config/target/iscsi/iqn.2003-01.org.linux-iscsi.osd1.x8664:sn.abcd1abcd2ab/tpgt_1/enable
+Directory: fixtures/sys/kernel/config/target/core/fileio_1/file_lio_1G
+Mode: 755
+# ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+Path: fixtures/sys/kernel/config/target/core/fileio_1/file_lio_1G/enable
```
Comment (pgier) [inline fixtures.ttar]: There seem to be a lot of changes to the fixtures that are unrelated to this PR. You might need to rebase on latest master and then update the fixtures again.

File: proc_fdinfo.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could not read %s: %s", f.Name(), err)
+	}
+
+	var text, pos, flags, mntid string
+	var inotify []InotifyInfo
+	rPos := regexp.MustCompile(`^pos:\s+(\d+)$`)
+	rFlags := regexp.MustCompile(`^flags:\s+(\d+)$`)
+	rMntID := regexp.MustCompile(`^mnt_id:\s+(\d+)$`)
+	rInotify := regexp.MustCompile(`^inotify`)
```
Comment (pgier) [inline proc_fdinfo.go]: These regexps should be moved to top level package variables so that they don't need to be compiled for each function call.

**Comment by pgier** [comment]
I think the fixtures issue is a separate issue and should be fixed by PR #185, so you can ignore that part.

File: proc_fdinfo.go (hunk 10 above)
Comment (pirxthepilot) [reply proc_fdinfo.go]: Sorry, I'm not yet familiar on how this is set. Do I just declare these at the top of the same file, after the imports?

Comment (pgier) [reply proc_fdinfo.go]: Yes, you can put them in a [`var ()` block](https://golang.org/doc/effective_go.html#variables) outside the function.  The location is mostly style preference, but either after the imports or just before the function where they are used are fine.

**Comment by pirxthepilot** [comment]
Updated!

**Comment by pgier** [comment]
Ug, sorry, the commit message lost the signoff when I did a squash merge.  I had to amend the commit and do a force push in #189.
